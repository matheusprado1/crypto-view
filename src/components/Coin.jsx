import { Link } from "react-router-dom"
import { TrendingDown, TrendingUp } from "../icons/icons"
import { formatBRLCurrency } from "../utils/formatCurrency"

const Coin = ({ coin }) => {
  return (
    <Link to={`/coin/${coin.id}`}>
      <div className="grid grid-cols-2 sm:grid-cols-3 p-2 rounded border-gray-200 border-b hover:bg-gray-800">
        <div className="flex item-center gap-2 w-full">
          <span className="font-semibold">{coin.market_cap_rank}.</span>
          <img className="w-8 h-8" src={coin.image} alt={coin.name} />
          <p className="font-semibold">{coin.name}</p>
          <span className="text-xs uppercase">({coin.symbol})</span>
        </div>
        <div className="hidden sm:block">
          <span className="text-sm font-semibold">Alta/Baixa 24h</span>
          <span className={`flex gap-2 font-semibold ${coin.price_change_percentage_24h < 0 ? "text-red-400" : "text-green-400"}`}>
            {coin.price_change_percentage_24h < 0 ? <TrendingDown /> : <TrendingUp />}
            {coin.price_change_percentage_24h}
            %</span>
        </div>
        <div className="ml-7">
          <p className="font-semibold">Pre√ßo Atual</p>
          <span>{formatBRLCurrency(coin.current_price)}</span>
        </div>
      </div>
    </Link>
  )
}

export default Coin
