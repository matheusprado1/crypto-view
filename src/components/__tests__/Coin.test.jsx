// eslint-disable-next-line no-unused-vars
import * as React from "react";
import TestRenderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { render, screen } from "@testing-library/react";
import Coin from "../Coin";

describe("Coin Component Tests", () => {
  const coin = {
    id: "1",
    market_cap_rank: 1,
    image: "https://example.com/image.png",
    name: "Bitcoin",
    symbol: "BTC",
    current_price: 50000,
  };

  // Teste para verificar se o componente renderiza corretamente quando a variação percentual do preço nas últimas 24 horas é positiva
  it("renders correctly when price_change_percentage_24h is positive", () => {
    const positiveCoin = { ...coin, price_change_percentage_24h: 5 };

    const tree = TestRenderer.create(
      <MemoryRouter>
        <Coin coin={positiveCoin} />
      </MemoryRouter>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  // Teste para verificar se o componente renderiza corretamente quando a variação percentual do preço nas últimas 24 horas é negativa
  it("renders correctly when price_change_percentage_24h is negative", () => {
    const negativeCoin = { ...coin, price_change_percentage_24h: -5 };

    const tree = TestRenderer.create(
      <MemoryRouter>
        <Coin coin={negativeCoin} />
      </MemoryRouter>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  // Teste para verificar se o nome e o símbolo da moeda são renderizados corretamente
  it("renders coin name and symbol correctly", () => {
    render(
      <MemoryRouter>
        <Coin coin={coin} />
      </MemoryRouter>
    );

    expect(screen.getByText(coin.name)).toBeInTheDocument();
    expect(screen.getByText(`(${coin.symbol})`)).toBeInTheDocument();
  });

  // Teste para verificar se a imagem da moeda é renderizada corretamente
  it("renders coin image correctly", () => {
    render(
      <MemoryRouter>
        <Coin coin={coin} />
      </MemoryRouter>
    );

    expect(screen.getByAltText(coin.name)).toHaveAttribute("src", coin.image);
  });

});
