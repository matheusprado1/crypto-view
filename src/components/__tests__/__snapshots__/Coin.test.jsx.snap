// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Coin > renders correctly 1`] = `
<a
  href="/coin/1"
  onClick={[Function]}
>
  <div
    className="grid grid-cols-2 sm:grid-cols-3 p-2 rounded border-gray-200 border-b hover:bg-gray-800"
  >
    <div
      className="flex item-center gap-2 w-full"
    >
      <span
        className="font-semibold"
      >
        1
        .
      </span>
      <img
        alt="Bitcoin"
        className="w-8 h-8"
        src="https://example.com/image.png"
      />
      <p
        className="font-semibold"
      >
        Bitcoin
      </p>
      <span
        className="text-xs uppercase"
      >
        (
        BTC
        )
      </span>
    </div>
    <div
      className="hidden sm:block"
    >
      <span
        className="text-sm font-semibold"
      >
        Alta/Baixa 24h
      </span>
      <span
        className="flex gap-2 font-semibold text-red-400"
      >
        <svg
          className="w-6 h-6"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M1.72 5.47a.75.75 0 0 1 1.06 0L9 11.69l3.756-3.756a.75.75 0 0 1 .985-.066 12.698 12.698 0 0 1 4.575 6.832l.308 1.149 2.277-3.943a.75.75 0 1 1 1.299.75l-3.182 5.51a.75.75 0 0 1-1.025.275l-5.511-3.181a.75.75 0 0 1 .75-1.3l3.943 2.277-.308-1.149a11.194 11.194 0 0 0-3.528-5.617l-3.809 3.81a.75.75 0 0 1-1.06 0L1.72 6.53a.75.75 0 0 1 0-1.061Z"
            fillRule="evenodd"
          />
        </svg>
        -5
        %
      </span>
    </div>
    <div
      className="ml-7"
    >
      <p
        className="font-semibold"
      >
        Preço Atual
      </p>
      <span>
        R$ 50.000,00
      </span>
    </div>
  </div>
</a>
`;

exports[`Coin > renders correctly when price_change_percentage_24h is negative 1`] = `
<a
  href="/coin/1"
  onClick={[Function]}
>
  <div
    className="grid grid-cols-2 sm:grid-cols-3 p-2 rounded border-gray-200 border-b hover:bg-gray-800"
  >
    <div
      className="flex item-center gap-2 w-full"
    >
      <span
        className="font-semibold"
      >
        1
        .
      </span>
      <img
        alt="Bitcoin"
        className="w-8 h-8"
        src="https://example.com/image.png"
      />
      <p
        className="font-semibold"
      >
        Bitcoin
      </p>
      <span
        className="text-xs uppercase"
      >
        (
        BTC
        )
      </span>
    </div>
    <div
      className="hidden sm:block"
    >
      <span
        className="text-sm font-semibold"
      >
        Alta/Baixa 24h
      </span>
      <span
        className="flex gap-2 font-semibold text-red-400"
      >
        <svg
          className="w-6 h-6"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M1.72 5.47a.75.75 0 0 1 1.06 0L9 11.69l3.756-3.756a.75.75 0 0 1 .985-.066 12.698 12.698 0 0 1 4.575 6.832l.308 1.149 2.277-3.943a.75.75 0 1 1 1.299.75l-3.182 5.51a.75.75 0 0 1-1.025.275l-5.511-3.181a.75.75 0 0 1 .75-1.3l3.943 2.277-.308-1.149a11.194 11.194 0 0 0-3.528-5.617l-3.809 3.81a.75.75 0 0 1-1.06 0L1.72 6.53a.75.75 0 0 1 0-1.061Z"
            fillRule="evenodd"
          />
        </svg>
        -5
        %
      </span>
    </div>
    <div
      className="ml-7"
    >
      <p
        className="font-semibold"
      >
        Preço Atual
      </p>
      <span>
        R$ 50.000,00
      </span>
    </div>
  </div>
</a>
`;

exports[`Coin > renders correctly when price_change_percentage_24h is positive 1`] = `
<a
  href="/coin/1"
  onClick={[Function]}
>
  <div
    className="grid grid-cols-2 sm:grid-cols-3 p-2 rounded border-gray-200 border-b hover:bg-gray-800"
  >
    <div
      className="flex item-center gap-2 w-full"
    >
      <span
        className="font-semibold"
      >
        1
        .
      </span>
      <img
        alt="Bitcoin"
        className="w-8 h-8"
        src="https://example.com/image.png"
      />
      <p
        className="font-semibold"
      >
        Bitcoin
      </p>
      <span
        className="text-xs uppercase"
      >
        (
        BTC
        )
      </span>
    </div>
    <div
      className="hidden sm:block"
    >
      <span
        className="text-sm font-semibold"
      >
        Alta/Baixa 24h
      </span>
      <span
        className="flex gap-2 font-semibold text-green-400"
      >
        <svg
          className="w-6 h-6"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M15.22 6.268a.75.75 0 0 1 .968-.431l5.942 2.28a.75.75 0 0 1 .431.97l-2.28 5.94a.75.75 0 1 1-1.4-.537l1.63-4.251-1.086.484a11.2 11.2 0 0 0-5.45 5.173.75.75 0 0 1-1.199.19L9 12.312l-6.22 6.22a.75.75 0 0 1-1.06-1.061l6.75-6.75a.75.75 0 0 1 1.06 0l3.606 3.606a12.695 12.695 0 0 1 5.68-4.974l1.086-.483-4.251-1.632a.75.75 0 0 1-.432-.97Z"
            fillRule="evenodd"
          />
        </svg>
        5
        %
      </span>
    </div>
    <div
      className="ml-7"
    >
      <p
        className="font-semibold"
      >
        Preço Atual
      </p>
      <span>
        R$ 50.000,00
      </span>
    </div>
  </div>
</a>
`;

exports[`Coin > renders correctly when price_change_percentage_24h is zero 1`] = `
<a
  href="/coin/1"
  onClick={[Function]}
>
  <div
    className="grid grid-cols-2 sm:grid-cols-3 p-2 rounded border-gray-200 border-b hover:bg-gray-800"
  >
    <div
      className="flex item-center gap-2 w-full"
    >
      <span
        className="font-semibold"
      >
        1
        .
      </span>
      <img
        alt="Bitcoin"
        className="w-8 h-8"
        src="https://example.com/image.png"
      />
      <p
        className="font-semibold"
      >
        Bitcoin
      </p>
      <span
        className="text-xs uppercase"
      >
        (
        BTC
        )
      </span>
    </div>
    <div
      className="hidden sm:block"
    >
      <span
        className="text-sm font-semibold"
      >
        Alta/Baixa 24h
      </span>
      <span
        className="flex gap-2 font-semibold text-green-400"
      >
        <svg
          className="w-6 h-6"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M15.22 6.268a.75.75 0 0 1 .968-.431l5.942 2.28a.75.75 0 0 1 .431.97l-2.28 5.94a.75.75 0 1 1-1.4-.537l1.63-4.251-1.086.484a11.2 11.2 0 0 0-5.45 5.173.75.75 0 0 1-1.199.19L9 12.312l-6.22 6.22a.75.75 0 0 1-1.06-1.061l6.75-6.75a.75.75 0 0 1 1.06 0l3.606 3.606a12.695 12.695 0 0 1 5.68-4.974l1.086-.483-4.251-1.632a.75.75 0 0 1-.432-.97Z"
            fillRule="evenodd"
          />
        </svg>
        0
        %
      </span>
    </div>
    <div
      className="ml-7"
    >
      <p
        className="font-semibold"
      >
        Preço Atual
      </p>
      <span>
        R$ 50.000,00
      </span>
    </div>
  </div>
</a>
`;

exports[`Coin Component Tests > renders correctly when price_change_percentage_24h is negative 1`] = `
<a
  href="/coin/1"
  onClick={[Function]}
>
  <div
    className="grid grid-cols-2 sm:grid-cols-3 p-2 rounded border-gray-200 border-b hover:bg-gray-800"
  >
    <div
      className="flex item-center gap-2 w-full"
    >
      <span
        className="font-semibold"
      >
        1
        .
      </span>
      <img
        alt="Bitcoin"
        className="w-8 h-8"
        src="https://example.com/image.png"
      />
      <p
        className="font-semibold"
      >
        Bitcoin
      </p>
      <span
        className="text-xs uppercase"
      >
        (
        BTC
        )
      </span>
    </div>
    <div
      className="hidden sm:block"
    >
      <span
        className="text-sm font-semibold"
      >
        Alta/Baixa 24h
      </span>
      <span
        className="flex gap-2 font-semibold text-red-400"
      >
        <svg
          className="w-6 h-6"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M1.72 5.47a.75.75 0 0 1 1.06 0L9 11.69l3.756-3.756a.75.75 0 0 1 .985-.066 12.698 12.698 0 0 1 4.575 6.832l.308 1.149 2.277-3.943a.75.75 0 1 1 1.299.75l-3.182 5.51a.75.75 0 0 1-1.025.275l-5.511-3.181a.75.75 0 0 1 .75-1.3l3.943 2.277-.308-1.149a11.194 11.194 0 0 0-3.528-5.617l-3.809 3.81a.75.75 0 0 1-1.06 0L1.72 6.53a.75.75 0 0 1 0-1.061Z"
            fillRule="evenodd"
          />
        </svg>
        -5
        %
      </span>
    </div>
    <div
      className="ml-7"
    >
      <p
        className="font-semibold"
      >
        Preço Atual
      </p>
      <span>
        R$ 50.000,00
      </span>
    </div>
  </div>
</a>
`;

exports[`Coin Component Tests > renders correctly when price_change_percentage_24h is positive 1`] = `
<a
  href="/coin/1"
  onClick={[Function]}
>
  <div
    className="grid grid-cols-2 sm:grid-cols-3 p-2 rounded border-gray-200 border-b hover:bg-gray-800"
  >
    <div
      className="flex item-center gap-2 w-full"
    >
      <span
        className="font-semibold"
      >
        1
        .
      </span>
      <img
        alt="Bitcoin"
        className="w-8 h-8"
        src="https://example.com/image.png"
      />
      <p
        className="font-semibold"
      >
        Bitcoin
      </p>
      <span
        className="text-xs uppercase"
      >
        (
        BTC
        )
      </span>
    </div>
    <div
      className="hidden sm:block"
    >
      <span
        className="text-sm font-semibold"
      >
        Alta/Baixa 24h
      </span>
      <span
        className="flex gap-2 font-semibold text-green-400"
      >
        <svg
          className="w-6 h-6"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M15.22 6.268a.75.75 0 0 1 .968-.431l5.942 2.28a.75.75 0 0 1 .431.97l-2.28 5.94a.75.75 0 1 1-1.4-.537l1.63-4.251-1.086.484a11.2 11.2 0 0 0-5.45 5.173.75.75 0 0 1-1.199.19L9 12.312l-6.22 6.22a.75.75 0 0 1-1.06-1.061l6.75-6.75a.75.75 0 0 1 1.06 0l3.606 3.606a12.695 12.695 0 0 1 5.68-4.974l1.086-.483-4.251-1.632a.75.75 0 0 1-.432-.97Z"
            fillRule="evenodd"
          />
        </svg>
        5
        %
      </span>
    </div>
    <div
      className="ml-7"
    >
      <p
        className="font-semibold"
      >
        Preço Atual
      </p>
      <span>
        R$ 50.000,00
      </span>
    </div>
  </div>
</a>
`;
